### PR TITLE
Review PR

### PR DESCRIPTION
**Pull request by dezuz**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,7 @@
+
```
Comment (Taniann) [inline .gitignore]: for future: that filemustn`'t be in your homework PR

File: src/com/mateacademy/collections/MyArrayList.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    @Override
+    protected Object clone() throws CloneNotSupportedException {
+        return super.clone();
+    }
+
+    @Override
+    public T get(int index) {
+        if (index > this.size ) {
```
Comment (Taniann) [inline src/com/mateacademy/collections/MyArrayList.java]: better dont write this when it isnt required. 
what if index < 0

File: src/com/mateacademy/collections/MyArrayList.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        if (size >= elements.length) {
+            elements = Arrays.copyOf(elements, elements.length * 2);
+        }
+
+        elements[size++] = t;
+    }
+
+    @Override
+    public void remove(int index) {
+        if (index > this.size ) {
```
Comment (Taniann) [inline src/com/mateacademy/collections/MyArrayList.java]: index also can be < 0

**Review by Taniann — CHANGES_REQUESTED** [review]
read comments

File: src/com/mateacademy/collections/MyArrayList.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+            elements = Arrays.copyOf(elements, elements.length * 2);
+        }
+
+        elements[size++] = t;
+    }
+
+    @Override
+    public void remove(int index) {
+        if (index < 0) {
+            System.out.println("Please write a valid index");
```
Comment (Taniann) [inline src/com/mateacademy/collections/MyArrayList.java]: better throw exception here with needed message

**Review by Taniann — APPROVED** [review]
good

File: .gitignore [hunk 5]
```diff
@@ -0,0 +1,4 @@
+.idea
+out/
+*.iml
+target/
```
Comment (Anderwerty) [inline .gitignore]: Add a blank line at the end of the file.

File: pom.xml [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+                <configuration>
+                    <source>8</source>
+                    <target>8</target>
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
+
+    <dependencies>
+        <!-- https://mvnrepository.com/artifact/junit/junit -->
```
Comment (Anderwerty) [inline pom.xml]: Redundant comment. Remove it.

File: pom.xml [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>4.12</version>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
+
+
+</project>
```
Comment (Anderwerty) [inline pom.xml]: Add blank line a the end of the file.

File: src/main/java/com/mateacademy/collections/MyArrayList.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        }
+
+        elements[size++] = t;
+    }
+
+    public void remove(int index) {
+        if (index < 0) {
+            throw new ArrayIndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+        else {
```
Comment (Anderwerty) [inline src/main/java/com/mateacademy/collections/MyArrayList.java]: You could remove else {

File: src/main/java/com/mateacademy/collections/MyArrayList.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+            if (index > this.size) {
+                throw new ArrayIndexOutOfBoundsException(index);
+            }
+        }
```
Comment (Anderwerty) [inline src/main/java/com/mateacademy/collections/MyArrayList.java:47]: remove }

File: src/main/java/com/mateacademy/collections/MyArrayList.java [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+        int numberMoved = size - index;
+        if(numberMoved > 0) {
+            System.arraycopy(elements, index, elements, index - 1,
```
Comment (Anderwerty) [inline src/main/java/com/mateacademy/collections/MyArrayList.java]: Why not in one line.

File: src/main/java/com/mateacademy/collections/MyArrayList.java [hunk 11]
```diff
@@ -0,0 +1,66 @@
+package com.mateacademy.collections;
+
+import java.util.Arrays;
+
+public class  MyArrayList <T> implements MyList<T>, Cloneable{
```
Comment (Anderwerty) [inline src/main/java/com/mateacademy/collections/MyArrayList.java:5]: Please add @Override annotation for all override methods.

File: src/main/java/com/mateacademy/collections/MyLinkedList.java [hunk 12]
```diff
@@ -0,0 +1,96 @@
+package com.mateacademy.collections;
+
+public class MyLinkedList<T> implements MyList<T> {
```
Comment (Anderwerty) [inline src/main/java/com/mateacademy/collections/MyLinkedList.java:3]: Add override annotation where it should be.

File: src/test/java/com/mateacademy/collections/MyArrayListTest.java [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+
+    @Test
+    public void testClearMethod() {
+        Integer expected = 0;
+        myArrayList.clear();
+        Integer actual = myArrayList.size();
+
+        assertEquals(expected, actual);
+    }
+}
```
Comment (Anderwerty) [inline src/test/java/com/mateacademy/collections/MyArrayListTest.java]: Add  a blank line at the end of the file.

File: src/test/java/com/mateacademy/collections/MyArrayListTest.java [hunk 14]
```diff
@@ -0,0 +1,73 @@
+package com.mateacademy.collections;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+
+public class MyArrayListTest {
```
Comment (Anderwerty) [inline src/test/java/com/mateacademy/collections/MyArrayListTest.java:10]: You should add tests for all public methods.

File: src/test/java/com/mateacademy/collections/MyLinkedListTest.java [hunk 15]
```diff
@@ -0,0 +1,63 @@
+package com.mateacademy.collections;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+
+public class MyLinkedListTest {
```
Comment (Anderwerty) [inline src/test/java/com/mateacademy/collections/MyLinkedListTest.java:10]: You should add tests for all public methods.

File: src/test/java/com/mateacademy/collections/MyLinkedListTest.java [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+    @Test
+    public void testClearMethod() {
+        Integer expected = 0;
+        myLinkedList.clear();
+        Integer actual = myLinkedList.size();
+
+        assertEquals(expected, actual);
+    }
+
+}
```
Comment (Anderwerty) [inline src/test/java/com/mateacademy/collections/MyLinkedListTest.java]: Add a blank line at the end of the file.

File: src/test/java/com/mateacademy/collections/MyStackTest.java [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+    public void testPopMethodToReturnCorrectSize() {
+        Integer expected = 3;
+        myStack.pop();
+        Integer actual = myStack.size();
+
+        assertEquals(expected, actual);
+    }
+
+
+}
```
Comment (Anderwerty) [inline src/test/java/com/mateacademy/collections/MyStackTest.java]: A blank line should be added.

**Review by Anderwerty — CHANGES_REQUESTED** [review]

File: src/main/java/com/mateacademy/collections/MyArrayList.java [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+            elements = Arrays.copyOf(elements, elements.length * 2);
+        }
+
+        elements[size++] = t;
+    }
+
+    @Override
+    public void remove(int index) {
+        if (index < 0) {
+            throw new ArrayIndexOutOfBoundsException("Index: " + index + ", Size: " + size);
```
Comment (Anderwerty) [inline src/main/java/com/mateacademy/collections/MyArrayList.java]: You should have 3 tests for this method.

File: src/main/java/com/mateacademy/collections/MyList.java [hunk 19]
```diff
@@ -0,0 +1,8 @@
+package com.mateacademy.collections;
+
+public interface MyList<T> {
```
Comment (Anderwerty) [inline src/main/java/com/mateacademy/collections/MyList.java:3]: Please format this code.

File: src/test/java/com/mateacademy/collections/MyStackTest.java [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+    @Test
+    public void testRemoveMethod() {
+        Integer expected = 4;
+        myStack.remove();
+        Integer actual = myStack.peek();
+
+        assertEquals(expected, actual);
+    }
+
+    @Test(expected = ArrayIndexOutOfBoundsException.class)
```
Comment (Anderwerty) [inline src/test/java/com/mateacademy/collections/MyStackTest.java]: Please also check exception message

File: src/test/java/com/mateacademy/collections/MyLinkedListTest.java [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+    public void testRemoveMethod() {
+        Integer expected = 9;
+        myLinkedList.remove(2);
+        Integer actual = myLinkedList.get(2);
+
+        assertEquals(expected, actual);
+    }
+
+    @Test(expected = ArrayIndexOutOfBoundsException.class)
+    public void testRemoveMethodToException() {
```
Comment (Anderwerty) [inline src/test/java/com/mateacademy/collections/MyLinkedListTest.java:57]: Please also check exception message

File: src/test/java/com/mateacademy/collections/MyLinkedListTest.java [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+    @Test
+    public void testAddMethod() {
+        Integer expected = 16;
+        myLinkedList.add(16);
+        Integer actual = myLinkedList.get(4);
+        assertEquals(expected, actual);
+    }
+
+    @Test(expected = ArrayIndexOutOfBoundsException.class)
+    public void testGetMethodToException() {
```
Comment (Anderwerty) [inline src/test/java/com/mateacademy/collections/MyLinkedListTest.java:41]: Please also check exception message

File: src/test/java/com/mateacademy/collections/MyArrayListTest.java [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+    public void testRemoveMethod() {
+        Object expected = elements[3];
+        myArrayList.remove(2);
+        Object actual = myArrayList.get(2);
+
+        assertEquals(expected, actual);
+    }
+
+    @Test(expected = ArrayIndexOutOfBoundsException.class)
+    public void testRemoveMethodToExceptions() {
```
Comment (Anderwerty) [inline src/test/java/com/mateacademy/collections/MyArrayListTest.java]: Please also check exception message

**Review by Anderwerty — CHANGES_REQUESTED** [review]

**Review by Anderwerty — APPROVED** [review]

**Pull request merged**